// m3_mini_projeto/src/api/fetchMentions.ts
import { get, getById, create, put, remove } from "./index.js";
import { MentionAPIRequest } from "./dto/index.js";

const ENDPOINT = "mentions";

/* Função para obter a lista de menções */
export async function getMentions(
  sort?: string,
  search?: string,
): Promise<MentionAPIRequest[]> {
  return get<MentionAPIRequest>(ENDPOINT, sort, search);
}

/* Função para obter uma menção por ID */
export async function getMentionById(
  id: number,
): Promise<MentionAPIRequest | null> {
  return getById<MentionAPIRequest>(ENDPOINT, id);
}

/* Função para criar uma nova menção */
export async function createMention(
  mention: Partial<MentionAPIRequest>,
): Promise<MentionAPIRequest | null> {
  return create<MentionAPIRequest>(ENDPOINT, mention);
}

/* Função para atualizar uma menção */
export async function updateMention(
  id: number,
  mention: Partial<MentionAPIRequest>,
): Promise<MentionAPIRequest | null> {
  return put<MentionAPIRequest>(ENDPOINT, id, mention);
}

/* Função para deletar uma menção */
export async function deleteMention(id: number): Promise<boolean> {
  return remove(ENDPOINT, id);
}
