// m3_mini_projeto/src/api/fetchTaskAttachments.ts
import { get, getById, create, put, remove } from "./index.js";
import { TaskAttachmentAPIRequest } from "./dto/index.js";

const ENDPOINT = "task_attachments";

/* Função para obter a lista de anexos de tarefas */
export async function getTaskAttachments(
  sort?: string,
  search?: string,
): Promise<TaskAttachmentAPIRequest[]> {
  return get<TaskAttachmentAPIRequest>(ENDPOINT, sort, search);
}

/* Função para obter um anexo de tarefa por ID */
export async function getTaskAttachmentById(
  id: number,
): Promise<TaskAttachmentAPIRequest | null> {
  return getById<TaskAttachmentAPIRequest>(ENDPOINT, id);
}

/* Função para criar um novo anexo de tarefa */
export async function createTaskAttachment(
  attachment: Partial<TaskAttachmentAPIRequest>,
): Promise<TaskAttachmentAPIRequest | null> {
  return create<TaskAttachmentAPIRequest>(ENDPOINT, attachment);
}

/* Função para atualizar um anexo de tarefa */
export async function updateTaskAttachment(
  id: number,
  attachment: Partial<TaskAttachmentAPIRequest>,
): Promise<TaskAttachmentAPIRequest | null> {
  return put<TaskAttachmentAPIRequest>(ENDPOINT, id, attachment);
}

/* Função para deletar um anexo de tarefa */
export async function deleteTaskAttachment(id: number): Promise<boolean> {
  return remove(ENDPOINT, id);
}
