// m3_mini_projeto/src/api/fetchPriorities.ts
import { get, getById, create, put, remove } from "./index.js";
import { PriorityAPIRequest } from "./dto/index.js";

const ENDPOINT = "priorities";

/* Função para obter todos as prioridades  */
export async function getPriorities(
  sort?: string,
  search?: string,
): Promise<PriorityAPIRequest[]> {
  return get<PriorityAPIRequest>(ENDPOINT, sort, search);
}

/* Função para obter uma prioridade pelo id */
export async function getPriorityById(
  id: number,
): Promise<PriorityAPIRequest | null> {
  return getById<PriorityAPIRequest>(ENDPOINT, id);
}

/* Função para criar uma nova prioridade  */
export async function createPriority(
  priority: Partial<PriorityAPIRequest>,
): Promise<PriorityAPIRequest | null> {
  return create<PriorityAPIRequest>(ENDPOINT, priority);
}

/* Função para editar ou atualizar os dados de uma prioridade  */
export async function updatePriority(
  id: number,
  priority: Partial<PriorityAPIRequest>,
): Promise<PriorityAPIRequest | null> {
  return put<PriorityAPIRequest>(ENDPOINT, id, priority);
}

/* Função para deletar uma prioridade */
export async function deletePriority(id: number): Promise<boolean> {
  return remove(ENDPOINT, id);
}
