// m3_mini_projeto/src/api/fetchProjectStatus.ts
import { get, getById, create, put, remove } from "./index.js";
import { ProjectStatusAPIRequest } from "./dto/index.js";

const ENDPOINT = "project_status";

/* Função para obter a lista de status de projeto */
export async function getProjectStatuses(
  sort?: string,
  search?: string,
): Promise<ProjectStatusAPIRequest[]> {
  return get<ProjectStatusAPIRequest>(ENDPOINT, sort, search);
}

/* Função para obter um status de projeto por ID */
export async function getProjectStatusById(
  id: number,
): Promise<ProjectStatusAPIRequest | null> {
  return getById<ProjectStatusAPIRequest>(ENDPOINT, id);
}

/* Função para criar um novo status de projeto */
export async function createProjectStatus(
  status: Partial<ProjectStatusAPIRequest>,
): Promise<ProjectStatusAPIRequest | null> {
  return create<ProjectStatusAPIRequest>(ENDPOINT, status);
}

/* Função para atualizar um status de projeto */
export async function updateProjectStatus(
  id: number,
  status: Partial<ProjectStatusAPIRequest>,
): Promise<ProjectStatusAPIRequest | null> {
  return put<ProjectStatusAPIRequest>(ENDPOINT, id, status);
}

/* Função para deletar um status de projeto */
export async function deleteProjectStatus(id: number): Promise<boolean> {
  return remove(ENDPOINT, id);
}
