// m3_mini_projeto/src/api/fetchNotifications.ts
import { get, getById, create, put, remove } from "./index.js";
import { NotificationAPIRequest } from "./dto/index.js";

const ENDPOINT = "notifications";

/* Função para obter a lista de notificações */
export async function getNotifications(
  sort?: string,
  search?: string,
): Promise<NotificationAPIRequest[]> {
  return get<NotificationAPIRequest>(ENDPOINT, sort, search);
}

/* Função para obter uma notificação específica por ID */
export async function getNotificationById(
  id: number,
): Promise<NotificationAPIRequest | null> {
  return getById<NotificationAPIRequest>(ENDPOINT, id);
}

/* Função para criar uma nova notificação */
export async function createNotification(
  notification: NotificationAPIRequest,
): Promise<NotificationAPIRequest | null> {
  return create<NotificationAPIRequest>(ENDPOINT, notification);
}

/* Função para atualizar uma notificação existente */
export async function updateNotification(
  id: number,
  notification: NotificationAPIRequest,
): Promise<NotificationAPIRequest | null> {
  return put<NotificationAPIRequest>(ENDPOINT, id, notification);
}

/* Função para excluir uma notificação por ID */
export async function deleteNotification(id: number): Promise<boolean> {
  return remove(ENDPOINT, id);
}
