// m3_mini_projeto/src/api/fetchTags.ts
import { get, getById, create, put, remove } from "./index.js";
import { BASE_URL } from "./utils/index.js";
import { TaskAPIRequest, TagAPIRequest } from "./dto/index.js";

const ENDPOINT = "tags";

/* Função para obter a lista de tags */
export async function getTags(
  sort?: string,
  search?: string,
): Promise<TagAPIRequest[]> {
  return get<TagAPIRequest>(ENDPOINT, sort, search);
}

/* Função para obter uma tag por ID */
export async function getTagById(id: number): Promise<TagAPIRequest | null> {
  return getById<TagAPIRequest>(ENDPOINT, id);
}

/* Função para criar uma nova tag */
export async function createTag(
  tag: Partial<TagAPIRequest>,
): Promise<TagAPIRequest | null> {
  return create<TagAPIRequest>(ENDPOINT, tag);
}

/* Função para atualizar uma tag */
export async function updateTag(
  id: number,
  tag: Partial<TagAPIRequest>,
): Promise<TagAPIRequest | null> {
  return put<TagAPIRequest>(ENDPOINT, id, tag);
}

/* Função para deletar uma tag */
export async function deleteTag(id: number): Promise<boolean> {
  return remove(ENDPOINT, id);
}

export async function getTasksByTag(tagId: number): Promise<TaskAPIRequest[]> {
  try {
    const res = await fetch(`${BASE_URL}${ENDPOINT}/${tagId}/tasks`);
    if (!res.ok) {
      throw new Error(`ERRO: Não foi possível obter tarefas da tag ${tagId} - ${res.status}`);
    }
    const data = await res.json();
    return data;
  } catch (error) {
    console.error("Erro ao obter tarefas da tag:", error);
    return [];
  }
}
