// m3_mini_projeto/src/api/fetchTaskStatusHistory.ts
import { get, getById, create, put, remove } from "./index.js";
import { TaskStatusHistoryAPIRequest } from "./dto/index.js";

const ENDPOINT = "task_status_history";

/* Função para obter a lista do histórico de status de tarefas */ 
export async function getTaskStatusHistories(
  sort?: string,
  search?: string,
): Promise<TaskStatusHistoryAPIRequest[]> {
  return get<TaskStatusHistoryAPIRequest>(ENDPOINT, sort, search);
}

/* Função para obter um registro de histórico de status por ID */
export async function getTaskStatusHistoryById(
  id: number,
): Promise<TaskStatusHistoryAPIRequest | null> {
  return getById<TaskStatusHistoryAPIRequest>(ENDPOINT, id);
}

/* Função para criar um novo registro de histórico de status */
export async function createTaskStatusHistory(
  history: Partial<TaskStatusHistoryAPIRequest>,
): Promise<TaskStatusHistoryAPIRequest | null> {
  return create<TaskStatusHistoryAPIRequest>(ENDPOINT, history);
}

/* Função para editar ohistórico de status */
export async function updateTaskStatusHistory(
  id: number,
  history: Partial<TaskStatusHistoryAPIRequest>,
): Promise<TaskStatusHistoryAPIRequest | null> {
  return put<TaskStatusHistoryAPIRequest>(ENDPOINT, id, history);
}

/* Função para deletar um registro de histórico de status */
export async function deleteTaskStatusHistory(id: number): Promise<boolean> {
  return remove(ENDPOINT, id);
}
