// m3_mini_projeto/src/api/fetchTeams.ts
import { get, getById, create, put, remove } from "./index.js";
import { TeamAPIRequest } from "./dto/index.js";

const ENDPOINT = "teams";

/* Função para obter a lista de equipes */
export async function getTeams(
  sort?: string,
  search?: string,
): Promise<TeamAPIRequest[]> {
  return get<TeamAPIRequest>(ENDPOINT, sort, search);
}

/* Função para obter uma equipe por ID */
export async function getTeamById(id: number): Promise<TeamAPIRequest | null> {
  return getById<TeamAPIRequest>(ENDPOINT, id);
}

/* Função para criar uma nova equipe */
export async function createTeam(
  team: Partial<TeamAPIRequest>,
): Promise<TeamAPIRequest | null> {
  return create<TeamAPIRequest>(ENDPOINT, team);
}

/* Função para atualizar uma equipe */
export async function updateTeam(
  id: number,
  team: Partial<TeamAPIRequest>,
): Promise<TeamAPIRequest | null> {
  return put<TeamAPIRequest>(ENDPOINT, id, team);
}

/* Função para deletar uma equipe */
export async function deleteTeam(id: number): Promise<boolean> {
  return remove(ENDPOINT, id);
}
