// m3_mini_projeto/src/api/fetchTeamMembers.ts
import { get, getById, create, put, remove } from "./index.js";
import { TeamMemberAPIRequest } from "./dto/index.js";

const ENDPOINT = "team_members";

/* Função para obter a lista de membros de equipe */
export async function getTeamMembers(
  sort?: string,
  search?: string,
): Promise<TeamMemberAPIRequest[]> {
  return get<TeamMemberAPIRequest>(ENDPOINT, sort, search);
}

/* Função para obter um membro de equipe por ID */
export async function getTeamMemberById(
  id: number,
): Promise<TeamMemberAPIRequest | null> {
  return getById<TeamMemberAPIRequest>(ENDPOINT, id);
}

/* Função para criar um novo membro de equipe */
export async function createTeamMember(
  member: Partial<TeamMemberAPIRequest>,
): Promise<TeamMemberAPIRequest | null> {
  return create<TeamMemberAPIRequest>(ENDPOINT, member);
}

/* Função para atualizar um membro de equipe */
export async function updateTeamMember(
  id: number,
  member: Partial<TeamMemberAPIRequest>,
): Promise<TeamMemberAPIRequest | null> {
  return put<TeamMemberAPIRequest>(ENDPOINT, id, member);
}

/* Função para deletar um membro de equipe */
export async function deleteTeamMember(id: number): Promise<boolean> {
  return remove(ENDPOINT, id);
}
