// m3_mini_projeto/src/api/fetchProjectPermissions.ts
import { get, getById, create, put, remove } from "./index.js";
import { ProjectPermissionAPIRequest } from "./dto/index.js";

const ENDPOINT = "project_permissions";

/* Função para obter a lista de permissões de projeto */
export async function getProjectPermissions(
  sort?: string,
  search?: string,
): Promise<ProjectPermissionAPIRequest[]> {
  return get<ProjectPermissionAPIRequest>(ENDPOINT, sort, search);
}

/* Função para obter uma permissão de projeto por ID */
export async function getProjectPermissionById(
  id: number,
): Promise<ProjectPermissionAPIRequest | null> {
  return getById<ProjectPermissionAPIRequest>(ENDPOINT, id);
}

/* Função para criar uma nova permissão de projeto */
export async function createProjectPermission(
  permission: Partial<ProjectPermissionAPIRequest>,
): Promise<ProjectPermissionAPIRequest | null> {
  return create<ProjectPermissionAPIRequest>(ENDPOINT, permission);
}

/* Função para atualizar uma permissão de projeto */
export async function updateProjectPermission(
  id: number,
  permission: Partial<ProjectPermissionAPIRequest>,
): Promise<ProjectPermissionAPIRequest | null> {
  return put<ProjectPermissionAPIRequest>(ENDPOINT, id, permission);
}

/* Função para deletar uma permissão de projeto */
export async function deleteProjectPermission(id: number): Promise<boolean> {
  return remove(ENDPOINT, id);
}
