// m3_mini_projeto/src/api/fetchTaskDependencies.ts
import { get, getById, create, put, remove } from "./index.js";
import { TaskDependencyAPIRequest } from "./dto/index.js";

const ENDPOINT = "task_dependencies";

/* Função para obter a lista de dependências de tarefas */
export async function getTaskDependencies(
  sort?: string,
  search?: string,
): Promise<TaskDependencyAPIRequest[]> {
  return get<TaskDependencyAPIRequest>(ENDPOINT, sort, search);
}

/* Função para obter uma dependência de tarefa por ID */
export async function getTaskDependencyById(
  id: number,
): Promise<TaskDependencyAPIRequest | null> {
  return getById<TaskDependencyAPIRequest>(ENDPOINT, id);
}

/* Função para criar uma nova dependência de tarefa */
export async function createTaskDependency(
  dependency: Partial<TaskDependencyAPIRequest>,
): Promise<TaskDependencyAPIRequest | null> {
  return create<TaskDependencyAPIRequest>(ENDPOINT, dependency);
}

/* Função para atualizar uma dependência de tarefa */
export async function updateTaskDependency(
  id: number,
  dependency: Partial<TaskDependencyAPIRequest>,
): Promise<TaskDependencyAPIRequest | null> {
  return put<TaskDependencyAPIRequest>(ENDPOINT, id, dependency);
}

/* Função para deletar uma dependência de tarefa */
export async function deleteTaskDependency(id: number): Promise<boolean> {
  return remove(ENDPOINT, id);
}
