// m3_mini_projeto/src/api/fetchCategories.ts
import { get, getById, create, put, remove } from "./index.js";
import { CategoryAPIRequest } from "./dto/index.js";

const ENDPOINT = "categories";

/* Função para obter a lista de categorias */
export async function getCategories(
  sort?: string,
  search?: string,
): Promise<CategoryAPIRequest[]> {
  return get<CategoryAPIRequest>(ENDPOINT, sort, search);
}

/* Função para obter uma categoria específica por ID */
export async function getCategoryById(id: number): Promise<CategoryAPIRequest | null> {
  return getById<CategoryAPIRequest>(ENDPOINT, id);
}

/* Função para criar uma nova categoria */
export async function createCategory(category: Partial<CategoryAPIRequest>): Promise<CategoryAPIRequest | null> {
  return create<CategoryAPIRequest>(ENDPOINT, category);
}

/* Função para atualizar uma categoria existente */
export async function updateCategory(
  id: number,
  category: Partial<CategoryAPIRequest>,
): Promise<CategoryAPIRequest | null> {
  return put<CategoryAPIRequest>(ENDPOINT, id, category);
}

/* Função para excluir uma categoria por ID */
export async function deleteCategory(id: number): Promise<boolean> {
  return remove(ENDPOINT, id);
}
