// m3_mini_projeto/src/api/fetchReminders.ts
import { get, getById, create, put, remove } from "./index.js";
import { ReminderAPIRequest } from "./dto/index.js";

const ENDPOINT = "reminders";

/* Função para obter a lista de lembretes */
export async function getReminders(
  sort?: string,
  search?: string,
): Promise<ReminderAPIRequest[]> {
  return get<ReminderAPIRequest>(ENDPOINT, sort, search);
}

/* Função para obter um lembrete por ID */
export async function getReminderById(
  id: number,
): Promise<ReminderAPIRequest | null> {
  return getById<ReminderAPIRequest>(ENDPOINT, id);
}

/* Função para criar um novo lembrete */
export async function createReminder(
  reminder: Partial<ReminderAPIRequest>,
): Promise<ReminderAPIRequest | null> {
  return create<ReminderAPIRequest>(ENDPOINT, reminder);
}

/* Função para atualizar um lembrete */
export async function updateReminder(
  id: number,
  reminder: Partial<ReminderAPIRequest>,
): Promise<ReminderAPIRequest | null> {
  return put<ReminderAPIRequest>(ENDPOINT, id, reminder);
}

/* Função para deletar um lembrete */
export async function deleteReminder(id: number): Promise<boolean> {
  return remove(ENDPOINT, id);
}
