// m3_mini_projeto/src/api/fetchSprints.ts
import { get, getById, create, put, remove } from "./index.js";
import { SprintAPIRequest } from "./dto/index.js";

const ENDPOINT = "sprints";

/* Função para obter a lista de sprints */
export async function getSprints(
  sort?: string,
  search?: string,
): Promise<SprintAPIRequest[]> {
  return get<SprintAPIRequest>(ENDPOINT, sort, search);
}

/* Função para obter um sprint por ID */
export async function getSprintById(
  id: number,
): Promise<SprintAPIRequest | null> {
  return getById<SprintAPIRequest>(ENDPOINT, id);
}

/* Função para criar um novo sprint */
export async function createSprint(
  sprint: Partial<SprintAPIRequest>,
): Promise<SprintAPIRequest | null> {
  return create<SprintAPIRequest>(ENDPOINT, sprint);
}

/* Função para atualizar um sprint */
export async function updateSprint(
  id: number,
  sprint: Partial<SprintAPIRequest>,
): Promise<SprintAPIRequest | null> {
  return put<SprintAPIRequest>(ENDPOINT, id, sprint);
}

/* Função para deletar um sprint */
export async function deleteSprint(id: number): Promise<boolean> {
  return remove(ENDPOINT, id);
}
