// m3_mini_projeto/src/api/fetchSprintTasks.ts
import { get, getById, create, put, remove } from "./index.js";
import { SprintTaskAPIRequest } from "./dto/index.js";

const ENDPOINT = "sprint_tasks";
/* Função para obter a lista de tarefas de sprint */
export async function getSprintTasks(
  sort?: string,
  search?: string,
): Promise<SprintTaskAPIRequest[]> {
  return get<SprintTaskAPIRequest>(ENDPOINT, sort, search);
}

/* Função para obter uma tarefa de sprint por ID */
export async function getSprintTaskById(
  id: number,
): Promise<SprintTaskAPIRequest | null> {
  return getById<SprintTaskAPIRequest>(ENDPOINT, id);
}

/* Função para criar uma nova tarefa de sprint */
export async function createSprintTask(
  sprintTask: Partial<SprintTaskAPIRequest>,
): Promise<SprintTaskAPIRequest | null> {
  return create<SprintTaskAPIRequest>(ENDPOINT, sprintTask);
}

/* Função para atualizar uma tarefa de sprint */
export async function updateSprintTask(
  id: number,
  sprintTask: Partial<SprintTaskAPIRequest>,
): Promise<SprintTaskAPIRequest | null> {
  return put<SprintTaskAPIRequest>(ENDPOINT, id, sprintTask);
}

/* Função para deletar uma tarefa de sprint */
export async function deleteSprintTask(id: number): Promise<boolean> {
  return remove(ENDPOINT, id);
}
