// m3_mini_projeto/dist/src/api/dto/mapperDTO.js
import { Project } from "../../projects/index.js";
import { UserClass } from "../../models/index.js";
import Notifications from "../../notifications/Notifications.js";
import { getStatus, getTaskStatusFromId, parseDate, } from "../../api/utils/index.js";
import { Task } from "../../tasks/Task.js";
/** Converter objeto plano para instância de UserClass */
export function mapToUserClass(data) {
    return new UserClass(data.id, data.name, data.email, data.phone, data.gender, getStatus(data.active), data.role);
}
/** Converter objeto plano para instância de Notifications */
export function mapToNotifications(data) {
    return new Notifications(data.id, data.title, data.message, data.is_read, data.sent_at);
}
/* Função auxiliar para mapear dados da API para instâncias de Project */
export function mapToProject(data) {
    return new Project(data.id, data.name, data.description, data.project_status_id, parseDate(data.start_date), parseDate(data.end_date_expected));
}
/** Converter instância de Project para objeto plano da API */
export function mapFromProject(project) {
    return {
        id: project.getId(),
        name: project.getName(),
        description: project.getDescription(),
        project_status_id: project.getProjectStatusId(),
        start_date: project.getStartDate().toISOString().split('T')[0],
        end_date_expected: project.getEndDateExpected().toISOString().split('T')[0],
    };
}
/** Converter objeto plano para instância de ITask (Task) */
export function mapToTask(data) {
    // Criar um projeto dummy com dados mínimos
    const project = new Project(data.project_id || 0, `Project ${data.project_id}`, "Project from API", 0, new Date(), new Date());
    // Criar uma categoria dummy com dados mínimos
    const category = {
        getId: () => data.category_id || 0,
        getName: () => `Category ${data.category_id}`,
    };
    // Criar tarefa usando a classe Task real
    const task = new Task(data.id, data.title, data.description || "", category, project);
    // Definir o status da tarefa
    task.setStatus(getTaskStatusFromId(data.status_id));
    // Se a tarefa foi marcada como concluída, definir isso
    if (data.completed_at) {
        task.markCompleted();
    }
    return task;
}
