// m3_mini_projeto/dist/src/ui/users/UserTaskDashboardUI.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { TaskStatus } from "../../tasks/TaskStatus.js";
import { renderModalEditTask } from "../modal/index.js";
import { getCardBorderColor, setCardBorderColor } from "../../helpers/index.js";
import { GenericDashboardUI } from "../dashboard/GenericDashboardUI.js";
import { UserService } from "../../services/index.js";
export class UserTaskDashboardUI {
    constructor(tasks) {
        this.usersCache = new Map();
        const taskColumns = [
            {
                id: TaskStatus.CREATED,
                label: "CREATED",
                filterFn: (task) => task.getStatus() === TaskStatus.CREATED,
            },
            {
                id: TaskStatus.ASSIGNED,
                label: "ASSIGNED",
                filterFn: (task) => task.getStatus() === TaskStatus.ASSIGNED,
            },
            {
                id: TaskStatus.BLOCKED,
                label: "BLOCKED",
                filterFn: (task) => task.getStatus() === TaskStatus.BLOCKED,
            },
            {
                id: TaskStatus.IN_PROGRESS,
                label: "IN_PROGRESS",
                filterFn: (task) => task.getStatus() === TaskStatus.IN_PROGRESS,
            },
            {
                id: TaskStatus.COMPLETED,
                label: "COMPLETED",
                filterFn: (task) => task.getStatus() === TaskStatus.COMPLETED,
            },
            {
                id: TaskStatus.ARCHIVED,
                label: "ARCHIVED",
                filterFn: (task) => task.getStatus() === TaskStatus.ARCHIVED,
            },
        ];
        this.dashboard = new GenericDashboardUI(tasks, {
            containerId: "dashBoardContainer",
            columns: taskColumns,
            cardRenderer: this.createTaskCard.bind(this),
            onCardClick: (task) => renderModalEditTask(task),
            itemId: (task) => task.getId(),
            itemGroupKey: (task) => task.getStatus(),
        });
        // Pré-carregar os nomes dos utilizadores
        this.loadUsersCache();
    }
    loadUsersCache() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const users = yield UserService.getUsers();
                users.forEach((user) => {
                    const id = typeof user.getId === 'function' ? user.getId() : user.id;
                    const name = typeof user.getName === 'function' ? user.getName() : user.name;
                    this.usersCache.set(id, name);
                });
            }
            catch (error) {
                console.error("Erro ao carregar nomes dos utilizadores:", error);
            }
        });
    }
    render() {
        return this.dashboard.render();
    }
    createTaskCard(task) {
        var _a, _b;
        const card = document.createElement("div");
        card.className = "task-card";
        const title = document.createElement("h3");
        title.className = "task-title";
        title.textContent = task.getTitle();
        const meta = document.createElement("div");
        meta.className = "task-meta";
        // Obter assignees da tarefa via TaskAssignees
        const assignees = ((_b = (_a = task).getAssignees) === null || _b === void 0 ? void 0 : _b.call(_a)) || [];
        const userSpan = document.createElement("span");
        userSpan.className = "task-user";
        if (assignees.length > 0) {
            // Obter nomes dos utilizadores atribuídos
            const userNames = assignees
                .map((assignee) => this.usersCache.get(assignee.user_id) || `User#${assignee.user_id}`)
                .join(", ");
            userSpan.textContent = userNames;
        }
        else {
            userSpan.textContent = "Não atribuído";
        }
        meta.appendChild(userSpan);
        const status = document.createElement("span");
        status.className = "task-status";
        status.textContent = task.getStatus();
        meta.appendChild(status);
        card.appendChild(title);
        card.appendChild(meta);
        // Aplicar cor do border baseada no status
        const borderColor = getCardBorderColor(task.getStatus());
        setCardBorderColor(card, borderColor);
        return card;
    }
}
