// m3_mini_projeto/dist/src/ui/users/UsersTaskDashboardUI.js
import { TaskStatus } from "../../tasks/TaskStatus.js";
import { createSection } from "../dom/CreatePage.js";
export class UsersTaskDashboardUI {
    constructor(user) {
        this.user = user;
        this.container = createSection("dashBoardContainer");
    }
    render() {
        this.container.innerHTML = "";
        this.container.classList.add("dash-board");
        const boardHTML = `
      <div class="dash-container">
        <div class="dash-column">
          <div class="column-header">
            <h2>CREATED</h2>
            <span class="column-count" data-status="CREATED">0</span>
          </div>
          <div class="tasks-container" data-status="CREATED"></div>
        </div>

        <div class="dash-column">
          <div class="column-header">
            <h2>ASSIGNED</h2>
            <span class="column-count" data-status="ASSIGNED">0</span>
          </div>
          <div class="tasks-container" data-status="ASSIGNED"></div>
        </div>

        <div class="dash-column">
          <div class="column-header">
            <h2>BLOCKED</h2>
            <span class="column-count" data-status="BLOCKED">0</span>
          </div>
          <div class="tasks-container" data-status="BLOCKED"></div>
        </div>

        <div class="dash-column">
          <div class="column-header">
            <h2>IN_PROGRESS</h2>
            <span class="column-count" data-status="IN_PROGRESS">0</span>
          </div>
          <div class="tasks-container" data-status="IN_PROGRESS"></div>
        </div>

        <div class="dash-column">
          <div class="column-header">
            <h2>COMPLETED</h2>
            <span class="column-count" data-status="COMPLETED">0</span>
          </div>
          <div class="tasks-container" data-status="COMPLETED"></div>
        </div>

           <div class="dash-column">
          <div class="column-header">
            <h2>ARCHIVED</h2>
            <span class="column-count" data-status="ARCHIVED">0</span>
          </div>
          <div class="tasks-container" data-status="ARCHIVED"></div>
        </div>
      </div>
    `;
        this.container.innerHTML = boardHTML;
        this.populateTasks();
        return this.container;
    }
    populateTasks() {
        const allUserTasks = [];
        // Agrupar tarefas por status
        const tasksByStatus = new Map();
        const statuses = Object.keys(TaskStatus).map((key) => TaskStatus[key]);
        statuses.forEach((status) => {
            tasksByStatus.set(status, []);
        });
        allUserTasks.forEach((task) => {
            const status = task.getStatus();
            const tasksForStatus = tasksByStatus.get(status) || [];
            tasksForStatus.push(task);
            tasksByStatus.set(status, tasksForStatus);
        });
        // Preencher cada coluna
        const statusesOrder = [
            TaskStatus.CREATED,
            TaskStatus.ASSIGNED,
            TaskStatus.BLOCKED,
            TaskStatus.IN_PROGRESS,
            TaskStatus.COMPLETED,
            TaskStatus.ARCHIVED,
        ];
        statusesOrder.forEach((status) => {
            const tasks = tasksByStatus.get(status) || [];
            const container = this.container.querySelector(`.tasks-container[data-status="${status}"]`);
            const countBadge = this.container.querySelector(`.column-count[data-status="${status}"]`);
            if (!container)
                return;
            // update badge only if it exists
            if (countBadge)
                countBadge.textContent = tasks.length.toString();
            // clear container then append tasks (safe for future partial updates)
            container.innerHTML = "";
            tasks.forEach((task) => {
                const taskCard = this.createTaskCard(task);
                container.appendChild(taskCard);
            });
        });
    }
    createTaskCard(task) {
        const card = document.createElement("div");
        card.className = "task-card";
        card.setAttribute("data-task-id", task.getId().toString());
        const title = document.createElement("h3");
        title.className = "task-title";
        title.textContent = task.getTitle();
        const meta = document.createElement("div");
        meta.className = "task-meta";
        const userTask = null;
        if (userTask) {
            const userSpan = document.createElement("span");
            userSpan.className = "task-user";
            userSpan.textContent = "";
            meta.appendChild(userSpan);
        }
        else {
            const userSpan = document.createElement("span");
            userSpan.className = "task-user";
            userSpan.textContent = this.user.getName();
            meta.appendChild(userSpan);
        }
        const status = document.createElement("span");
        status.className = "task-status";
        status.textContent = task.getStatus();
        meta.appendChild(status);
        card.appendChild(title);
        card.appendChild(meta);
        card.addEventListener("click", () => {
            this.handleTaskClick(task);
        });
        return card;
    }
    handleTaskClick(task) {
        const currentStatus = task.getStatus();
        const nextStatus = this.getNextStatus(currentStatus);
        if (nextStatus) {
            task.moveTo(nextStatus);
            this.render();
        }
    }
    getNextStatus(currentStatus) {
        const statusFlow = [
            TaskStatus.CREATED,
            TaskStatus.ASSIGNED,
            TaskStatus.IN_PROGRESS,
            TaskStatus.COMPLETED,
            TaskStatus.ARCHIVED,
        ];
        const currentIndex = statusFlow.indexOf(currentStatus);
        if (currentIndex < statusFlow.length - 1) {
            return statusFlow[currentIndex + 1];
        }
        return null;
    }
}
