// m3_mini_projeto/dist/src/api/fetchUsers.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { get, getById, create, put, patch, remove } from "./index.js";
import { BASE_URL } from "./utils/index.js";
import { mapToUserClass, mapToNotifications, } from "./dto/index.js";
const ENDPOINT = "users";
/* Função para obter a lista de utilizadores */
export function apiGetUsers(sort, search) {
    return __awaiter(this, void 0, void 0, function* () {
        const data = yield get(ENDPOINT, sort, search);
        return data.map(mapToUserClass);
    });
}
/* Função para obter um utilizador por ID */
export function apiGetUserById(id) {
    return __awaiter(this, void 0, void 0, function* () {
        const data = yield getById(ENDPOINT, id);
        return data ? mapToUserClass(data) : null;
    });
}
/* Função para obter estatísticas de utilizadores */
export function apiGetUserStats() {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const res = yield fetch(`${BASE_URL}${ENDPOINT}/stats`);
            if (!res.ok) {
                throw new Error("ERRO: Não foi possível obter estatísticas " + res.status);
            }
            const data = yield res.json();
            return data;
        }
        catch (error) {
            console.error("Erro ao obter estatísticas:", error);
            return null;
        }
    });
}
/* Função para obter notificações não lidas do utilizador */
export function apiGetUnreadNotifications(userId) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const res = yield fetch(`${BASE_URL}${ENDPOINT}/${userId}/notifications/unread`);
            if (!res.ok) {
                throw new Error("ERRO: Não foi possível obter notificações não lidas " + res.status);
            }
            const data = yield res.json();
            const notifications = Array.isArray(data)
                ? data.map(mapToNotifications)
                : [mapToNotifications(data)];
            return notifications;
        }
        catch (error) {
            console.error("Erro ao obter notificações não lidas:", error);
            return [];
        }
    });
}
/* Função para obter todas as notificações do utilizador */
export function apiGetNotificationsByUser(userId) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const res = yield fetch(`${BASE_URL}${ENDPOINT}/${userId}/notifications`);
            if (!res.ok) {
                throw new Error("ERRO: Não foi possível obter notificações " + res.status);
            }
            const data = yield res.json();
            const notifications = Array.isArray(data)
                ? data.map(mapToNotifications)
                : [mapToNotifications(data)];
            return notifications;
        }
        catch (error) {
            console.error("Erro ao obter notificações:", error);
            return [];
        }
    });
}
/* ======================== POST ======================== */
/* Função para criar um novo utilizador */
export function apiCreateUser(userData) {
    return __awaiter(this, void 0, void 0, function* () {
        const data = yield create(ENDPOINT, userData);
        return data ? mapToUserClass(data) : null;
    });
}
/* Função para atualizar um utilizador */
export function apiUpdateUser(userId, userData) {
    return __awaiter(this, void 0, void 0, function* () {
        const data = yield put(ENDPOINT, userId, userData);
        return data ? mapToUserClass(data) : null;
    });
}
/* Função para ativar/desativar um utilizador */
export function apiToggleUserActive(userId, active) {
    return __awaiter(this, void 0, void 0, function* () {
        const data = yield patch(ENDPOINT, userId, {
            active: active ? 1 : 0,
        });
        return data ? mapToUserClass(data) : null;
    });
}
/* Função para marcar uma notificação como lida */
export function apiMarkNotificationAsRead(userId, notificationId) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const res = yield fetch(`${BASE_URL}${ENDPOINT}/${userId}/notifications/${notificationId}`, {
                method: "PATCH",
                headers: {
                    "Content-Type": "application/json",
                },
            });
            if (!res.ok) {
                throw new Error("ERRO: Não foi possível marcar notificação como lida " + res.status);
            }
            const data = yield res.json();
            return data;
        }
        catch (error) {
            console.error("Erro ao marcar notificação como lida:", error);
            return null;
        }
    });
}
/* Função para deletar um utilizador */
export function apiDeleteUser(userId) {
    return __awaiter(this, void 0, void 0, function* () {
        return remove(ENDPOINT, userId);
    });
}
