// m3_mini_projeto/dist/src/api/fetchTags.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { get, getById, create, put, remove } from "./index.js";
import { BASE_URL } from "./utils/index.js";
const ENDPOINT = "tags";
/* Função para obter a lista de tags */
export function getTags(sort, search) {
    return __awaiter(this, void 0, void 0, function* () {
        return get(ENDPOINT, sort, search);
    });
}
/* Função para obter uma tag por ID */
export function getTagById(id) {
    return __awaiter(this, void 0, void 0, function* () {
        return getById(ENDPOINT, id);
    });
}
/* Função para criar uma nova tag */
export function createTag(tag) {
    return __awaiter(this, void 0, void 0, function* () {
        return create(ENDPOINT, tag);
    });
}
/* Função para atualizar uma tag */
export function updateTag(id, tag) {
    return __awaiter(this, void 0, void 0, function* () {
        return put(ENDPOINT, id, tag);
    });
}
/* Função para deletar uma tag */
export function deleteTag(id) {
    return __awaiter(this, void 0, void 0, function* () {
        return remove(ENDPOINT, id);
    });
}
export function getTasksByTag(tagId) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const res = yield fetch(`${BASE_URL}${ENDPOINT}/${tagId}/tasks`);
            if (!res.ok) {
                throw new Error(`ERRO: Não foi possível obter tarefas da tag ${tagId} - ${res.status}`);
            }
            const data = yield res.json();
            return data;
        }
        catch (error) {
            console.error("Erro ao obter tarefas da tag:", error);
            return [];
        }
    });
}
