// m3_mini_projeto/dist/src/api/fetchTasks.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { get, getById, create, put, patch, remove } from "./index.js";
import { BASE_URL } from "./utils/index.js";
const ENDPOINT = "tasks";
/* ======================== GET ======================== */
/* Função para obter a lista de tarefas */
export function getTasks(sort, search) {
    return __awaiter(this, void 0, void 0, function* () {
        return get(ENDPOINT, sort, search);
    });
}
/* Função para obter tarefas de um projeto específico */
export function getTasksByProject(projectId, sort, search) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            let url = `${BASE_URL}${ENDPOINT}/project/${projectId}`;
            if (sort || search) {
                const params = new URLSearchParams();
                if (sort)
                    params.append("sort", sort);
                if (search)
                    params.append("search", search);
                url += `?${params.toString()}`;
            }
            const res = yield fetch(url);
            if (!res.ok) {
                throw new Error(`HTTP ${res.status}: Não foi possível obter tarefas do projeto`);
            }
            const data = yield res.json();
            return data;
        }
        catch (error) {
            console.error("❌ Erro ao obter tarefas do projeto:", error);
            throw error;
        }
    });
}
/* Função para obter uma tarefa por ID */
export function getTaskById(taskId) {
    return __awaiter(this, void 0, void 0, function* () {
        return getById(ENDPOINT, taskId);
    });
}
/* Função para obter estatísticas de tarefas */
export function getTaskStats() {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const res = yield fetch(`${BASE_URL}${ENDPOINT}/stats`);
            if (!res.ok) {
                throw new Error("ERRO: Não foi possível obter estatísticas " + res.status);
            }
            const data = yield res.json();
            return data;
        }
        catch (error) {
            console.error("Erro ao obter estatísticas:", error);
            return null;
        }
    });
}
/* Função para obter tags de uma tarefa */
export function getTaskTags(taskId) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const res = yield fetch(`${BASE_URL}${ENDPOINT}/${taskId}/tags`);
            if (!res.ok) {
                throw new Error("ERRO: Não foi possível obter tags " + res.status);
            }
            const data = yield res.json();
            return data;
        }
        catch (error) {
            console.error("Erro ao obter tags:", error);
            return [];
        }
    });
}
/* Função para obter comentários de uma tarefa */
export function getTaskComments(taskId) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const res = yield fetch(`${BASE_URL}${ENDPOINT}/${taskId}/comments`);
            if (!res.ok) {
                throw new Error("ERRO: Não foi possível obter comentários " + res.status);
            }
            const data = yield res.json();
            return data;
        }
        catch (error) {
            console.error("Erro ao obter comentários:", error);
            return [];
        }
    });
}
/* Função para criar uma nova tarefa */
export function createTask(taskData) {
    return __awaiter(this, void 0, void 0, function* () {
        return create(ENDPOINT, taskData);
    });
}
/* Função para adicionar uma tag a uma tarefa */
export function addTagToTask(taskId, tagData) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const res = yield fetch(`${BASE_URL}${ENDPOINT}/${taskId}/tags`, {
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                },
                body: JSON.stringify(tagData),
            });
            if (!res.ok) {
                throw new Error("ERRO: Não foi possível adicionar tag " + res.status);
            }
            const data = yield res.json();
            return data;
        }
        catch (error) {
            console.error("Erro ao adicionar tag:", error);
            return null;
        }
    });
}
/* Função para criar um comentário em uma tarefa */
export function createTaskComment(taskId, commentData) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const res = yield fetch(`${BASE_URL}${ENDPOINT}/${taskId}/comments`, {
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                },
                body: JSON.stringify(commentData),
            });
            if (!res.ok) {
                throw new Error("ERRO: Não foi possível criar comentário " + res.status);
            }
            const data = yield res.json();
            return data;
        }
        catch (error) {
            console.error("Erro ao criar comentário:", error);
            return null;
        }
    });
}
/* Função para atualizar uma tarefa */
export function updateTask(taskId, taskData) {
    return __awaiter(this, void 0, void 0, function* () {
        return put(ENDPOINT, taskId, taskData);
    });
}
/* Função para atualizar um comentário */
export function updateTaskComment(taskId, commentId, commentData) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const res = yield fetch(`${BASE_URL}${ENDPOINT}/${taskId}/comments/${commentId}`, {
                method: "PUT",
                headers: {
                    "Content-Type": "application/json",
                },
                body: JSON.stringify(commentData),
            });
            if (!res.ok) {
                throw new Error("ERRO: Não foi possível atualizar comentário " + res.status);
            }
            const data = yield res.json();
            return data;
        }
        catch (error) {
            console.error("Erro ao atualizar comentário:", error);
            return null;
        }
    });
}
/* Função para atualizar o status de uma tarefa */
export function changeTaskStatus(taskId, statusId) {
    return __awaiter(this, void 0, void 0, function* () {
        return patch(ENDPOINT, taskId, { status_id: statusId });
    });
}
/* Função para resolver um comentário */
export function resolveTaskComment(taskId, commentId) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const res = yield fetch(`${BASE_URL}${ENDPOINT}/${taskId}/comments/${commentId}`, {
                method: "PATCH",
                headers: {
                    "Content-Type": "application/json",
                },
            });
            if (!res.ok) {
                throw new Error("ERRO: Não foi possível resolver comentário " + res.status);
            }
            const data = yield res.json();
            return data;
        }
        catch (error) {
            console.error("Erro ao resolver comentário:", error);
            return null;
        }
    });
}
/* Função para deletar uma tarefa */
export function deleteTask(taskId) {
    return __awaiter(this, void 0, void 0, function* () {
        return remove(ENDPOINT, taskId);
    });
}
/* Função para remover uma tag de uma tarefa */
export function removeTagFromTask(taskId, tagId) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const res = yield fetch(`${BASE_URL}${ENDPOINT}/${taskId}/tags/${tagId}`, {
                method: "DELETE",
            });
            if (!res.ok) {
                throw new Error("ERRO: Não foi possível remover tag " + res.status);
            }
            return true;
        }
        catch (error) {
            console.error("Erro ao remover tag:", error);
            return false;
        }
    });
}
/* Função para deletar um comentário */
export function deleteTaskComment(taskId, commentId) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const res = yield fetch(`${BASE_URL}${ENDPOINT}/${taskId}/comments/${commentId}`, {
                method: "DELETE",
            });
            if (!res.ok) {
                throw new Error("ERRO: Não foi possível deletar comentário " + res.status);
            }
            return true;
        }
        catch (error) {
            console.error("Erro ao deletar comentário:", error);
            return false;
        }
    });
}
