// m3_mini_projeto/dist/src/ui/projects/ProjectsCountersUI.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { ProjectService } from "../../services/index.js";
export function showProjectsCounters(type, projects) {
    return __awaiter(this, void 0, void 0, function* () {
        if ((type === "filtrados" || type === "ativos" || type === "concluidos") && projects) {
            countAllProjects("#allProjectsCounter", projects.length);
            if (type === "ativos") {
                const activeCount = projects.filter((p) => p.getStatus() === "Ativo").length;
                countActiveProjects("#activeProjectsCounter", activeCount);
                countFinishedProjects("#finishedProjectsCounter", 0);
                countInDevelopmentProjects("#inDevelopmentProjectsCounter", 0);
            }
            else if (type === "concluidos") {
                const finishedCount = projects.filter((p) => p.getStatus() === "Terminado").length;
                countFinishedProjects("#finishedProjectsCounter", finishedCount);
                countActiveProjects("#activeProjectsCounter", 0);
                countInDevelopmentProjects("#inDevelopmentProjectsCounter", 0);
            }
            else {
                countActiveProjects("#activeProjectsCounter", projects.length);
                countFinishedProjects("#finishedProjectsCounter", projects.length);
                countInDevelopmentProjects("#inDevelopmentProjectsCounter", projects.length);
            }
            countFilterProjects("#filterProjectsCounter", type, projects.length);
            yield countActivePercentage("#activeProjectsPercentageCounter", type);
        }
        else {
            yield countAllProjects("#allProjectsCounter");
            yield countActiveProjects("#activeProjectsCounter");
            yield countFinishedProjects("#finishedProjectsCounter");
            yield countInDevelopmentProjects("#inDevelopmentProjectsCounter");
            countFilterProjects("#filterProjectsCounter", type);
            yield countActivePercentage("#activeProjectsPercentageCounter", type);
        }
    });
}
/* Contador de projetos ativos */
function countActiveProjects(id, overrideValue) {
    return __awaiter(this, void 0, void 0, function* () {
        const section = document.querySelector(`${id}`);
        if (overrideValue !== undefined) {
            if (section) {
                section.textContent = `${overrideValue}`;
            }
            return;
        }
        const projects = yield ProjectService.getProjects();
        const activeCount = projects.filter((p) => p.getStatus() === "Ativo").length;
        if (section) {
            section.textContent = `${activeCount}`;
        }
        else {
            console.warn(`Elemento ${id} não foi encontrado no DOM.`);
        }
    });
}
/* Contador de projetos concluídos */
function countFinishedProjects(id, overrideValue) {
    return __awaiter(this, void 0, void 0, function* () {
        const section = document.querySelector(`${id}`);
        if (overrideValue !== undefined) {
            if (section) {
                section.textContent = `${overrideValue}`;
            }
            return;
        }
        const projects = yield ProjectService.getProjects();
        const finishedCount = projects.filter((p) => p.getStatus() === "Terminado").length;
        if (section) {
            section.textContent = `${finishedCount}`;
        }
        else {
            console.warn(`Elemento ${id} não foi encontrado no DOM.`);
        }
    });
}
/* Contador de projetos em desenvolvimento */
function countInDevelopmentProjects(id, overrideValue) {
    return __awaiter(this, void 0, void 0, function* () {
        const section = document.querySelector(`${id}`);
        if (overrideValue !== undefined) {
            if (section) {
                section.textContent = `${overrideValue}`;
            }
            return;
        }
        const projects = yield ProjectService.getProjects();
        const inDevCount = projects.filter((p) => p.getStatus() === "Em Desenvolvimento").length;
        if (section) {
            section.textContent = `${inDevCount}`;
        }
        else {
            console.warn(`Elemento ${id} não foi encontrado no DOM.`);
        }
    });
}
/* Contador de projetos filtrados */
function countFilterProjects(id, type, count) {
    const section = document.querySelector(`${id}`);
    if (section) {
        if (count !== undefined) {
            section.textContent = `${count}`;
        }
        else if (type === "projectFiltered" && section.textContent !== "") {
            section.textContent = `${0}`;
        }
        else {
            section.textContent = "0";
        }
    }
    else {
        console.warn(`Elemento ${id} não foi encontrado no DOM.`);
    }
}
/* Contador de todos os projetos */
function countAllProjects(id, overrideValue) {
    return __awaiter(this, void 0, void 0, function* () {
        const section = document.querySelector(`${id}`);
        if (overrideValue !== undefined) {
            if (section) {
                section.textContent = `${overrideValue}`;
            }
            return;
        }
        const projects = yield ProjectService.getProjects();
        if (section) {
            section.textContent = `${projects.length}`;
        }
        else {
            console.warn(`Elemento ${id} não foi encontrado no DOM.`);
        }
    });
}
/* Percentagem de projetos ativos */
function countActivePercentage(id, type) {
    return __awaiter(this, void 0, void 0, function* () {
        const section = document.querySelector(`${id}`);
        const projects = yield ProjectService.getProjects();
        const total = projects.length;
        if (section && total > 0) {
            let percentage = 0;
            if (type === "concluidos") {
                const finishedCount = projects.filter((p) => p.getStatus() === "Terminado").length;
                percentage = Math.round((finishedCount / total) * 100);
            }
            else {
                const activeCount = projects.filter((p) => p.getStatus() === "Ativo").length;
                percentage = Math.round((activeCount / total) * 100);
            }
            section.textContent = `${percentage}%`;
            changeImageAndFigCaption(type);
        }
        else if (section) {
            section.textContent = "0%";
        }
        else {
            console.warn(`Elemento ${id} não foi encontrado no DOM.`);
        }
    });
}
function changeImageAndFigCaption(type) {
    if (type) {
        const projectsPercentageCaption = document.querySelector("#projectsPercentageCaption");
        const activeProjectsPercentageBtn = document.querySelector("#activeProjectsPercentageBtn");
        if (projectsPercentageCaption && activeProjectsPercentageBtn) {
            switch (type) {
                case "concluidos":
                    activeProjectsPercentageBtn.title =
                        "Mostrar percentagem de projetos concluídos";
                    activeProjectsPercentageBtn.src = "./src/assets/grafico.png";
                    projectsPercentageCaption.textContent = "concluídos %";
                    break;
                case "projetos":
                case "ativos":
                    activeProjectsPercentageBtn.title =
                        "Mostrar percentagem de projetos ativos";
                    activeProjectsPercentageBtn.src = "./src/assets/percentagem.png";
                    projectsPercentageCaption.textContent = "ativos %";
                    break;
                default:
            }
        }
        else {
            console.warn(`Elemento projectsPercentageCaption não foi encontrado no DOM.`);
        }
    }
}
