// m3_mini_projeto/src/api/dto/typesDTO.ts
import { UserRole } from "../../security/UserRole.js";

/* Type para representar utilizador vindo da API */
export type UserAPIRequest = {
  id: number;
  name: string;
  email: string;
  phone: number;
  gender: string;
  active: number;
  role?: UserRole;
};

/* Type para representar estatísticas de utlizador vindo da API */
export type UserStatsAPIRequest = {
  totalUsers: number;
  activeUsers: number;
  inactiveUsers: number;
  activePercentage: string;
  inactivePercentage: string;
};

/* Type para representar notificação vindo da API */
export type NotificationAPIRequest = {
  id: number;
  title: string;
  message: string;
  is_read: number;
  sent_at: string;
};

/* Type para representar tarefa vindo da API */
export type TaskAPIRequest = {
  id: number;
  title: string;
  description?: string;
  status_id?: number;
  priority_id?: number;
  category_id?: number;
  assigned_to?: number;
  project_id: number;
  sprint_id?: number;
  start_date?: string;
  due_date?: string;
  completed_at?: string;
  created_at?: string;
  updated_at?: string;
};

/* Type para representar estatísticas de tarefas vindo da API */
export type TaskStatsAPIRequest = {
  totalTasks: number;
  completedTasks: number;
  pendingTasks: number;
  completedPercentage: string;
};

/** Type para representar comentário de tarefa vindo da API */
export type TaskCommentAPIRequest = {
  id: number;
  task_id: number;
  user_id: number;
  content: string;
  created_at?: string;
  updated_at?: string;
};

/** Type para representar tag vindo da API */
export type TagAPIRequest = {
  id: number;
  name: string;
  color?: string;
  description?: string;
};

/** Type para representar projeto vindo da API */
export type ProjectAPIRequest = {
  id: number;
  name: string;
  description?: string;
  project_status_id?: number;
  start_date?: string;
  end_date_expected?: string;
  created_at?: string;
  updated_at?: string;
};

/** Type para representar status de projeto vindo da API */
export type ProjectStatusAPIRequest = {
  id: number;
  name: string;
  description?: string;
};

/** Type para representar permissão de projeto vindo da API */
export type ProjectPermissionAPIRequest = {
  id: number;
  project_id: number;
  user_id: number;
  role: string;
};

/** Type para representar equipe vindo da API */
export type TeamAPIRequest = {
  id: number;
  name: string;
  description?: string;
  created_at?: string;
  updated_at?: string;
};

/** Type para representar membro de equipe vindo da API */
export type TeamMemberAPIRequest = {
  id: number;
  team_id: number;
  user_id: number;
  role?: string;
};

/** Type para representar sprint vindo da API */
export type SprintAPIRequest = {
  id: number;
  project_id: number;
  name: string;
  description?: string;
  start_date?: string;
  end_date?: string;
  status?: string;
  created_at?: string;
  updated_at?: string;
};

/** Type para representar tarefa de sprint vindo da API */
export type SprintTaskAPIRequest = {
  id: number;
  sprint_id: number;
  task_id: number;
};

/** Type para representar prioridade vindo da API */
export type PriorityAPIRequest = {
  id: number;
  name: string;
  level?: number;
  color?: string;
};

/** Type para representar status de tarefa vindo da API */
export type TaskStatusAPIRequest = {
  id: number;
  name: string;
  color?: string;
  description?: string;
};

/** Type para representar histórico de status vindo da API */
export type TaskStatusHistoryAPIRequest = {
  id: number;
  task_id: number;
  status_id: number;
  changed_by: number;
  changed_at: string;
  previous_status_id?: number;
};

/** Type para representar anexo de tarefa vindo da API */
export type TaskAttachmentAPIRequest = {
  id: number;
  task_id: number;
  file_path: string;
  file_name: string;
  uploaded_by: number;
  uploaded_at?: string;
};

/** Type para representar dependência de tarefa vindo da API */
export type TaskDependencyAPIRequest = {
  id: number;
  task_id: number;
  depends_on_task_id: number;
  dependency_type?: string;
};

/** Type para representar votação de tarefa vindo da API */
export type TaskVoteAPIRequest = {
  id: number;
  task_id: number;
  user_id: number;
  vote_type: string;
};

/** Type para representar menção vindo da API */
export type MentionAPIRequest = {
  id: number;
  user_id: number;
  task_id?: number;
  comment_id?: number;
  mentioned_by: number;
  created_at?: string;
};

/** Type para representar lembrete vindo da API */
export type ReminderAPIRequest = {
  id: number;
  task_id: number;
  user_id: number;
  reminder_date: string;
  message?: string;
  is_sent: number;
};

/** Type para representar log de tempo vindo da API */
export type TimeLogAPIRequest = {
  id: number;
  task_id: number;
  user_id: number;
  hours: number;
  description?: string;
  log_date: string;
};

/** Type para representar tarefa e tag vindo da API */
export type TagTaskAPIRequest = {
  id: number;
  task_id: number;
  tag_id: number;
};

/** Type para representar tarefa favorita vindo da API */
export type FavoriteTaskAPIRequest = {
  id: number;
  task_id: number;
  user_id: number;
};

/** Type para representar atribuição de tarefa vindo da API */
export type TaskAssigneeAPIRequest = {
  id: number;
  task_id: number;
  user_id: number;
  assigned_by: number;
  assigned_at?: string;
};

/** Type para representar categoria vindo da API */
export type CategoryAPIRequest = {
  id: number;
  name: string;
  description?: string;
};

/* Type para representar ranking de horas mais reais vindo da API */
export type RankingMoreHoursAPIRequest = {
  projeto: string;
  utilizador: string;
  total_horas_reais: number;
  ranking: number;
};

/* Mapeia a resposta da API para ranking de horas de aumento */
export type RankingIncreaseHoursAPIRequest = {
  utilizador: string;
  data_dia: string;
  horas_dia: number;
  horas_dia_anterior: number;
  ranking_do_dia: number;
};

/* Type para representar ranking de horas acima da média vindo da API */
export type RankingAboveAverageAPIRequest = {
  projeto: string;
  total_horas_projeto: number;
  media_geral_sistema: number;
};
