// m3_mini_projeto/dist/src/ui/tasks/TaskDashboardUI.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { createTaskCard } from "./TaskCardUI.js";
import { ProjectService, TaskService, UserService } from "../../services/index.js";
export class TaskDashboardUI {
    constructor() {
        this.projects = [];
        this.tasksByProject = new Map();
        this.userMap = new Map();
        this.container = document.createElement("div");
        this.container.id = "taskDashboardContainer";
        this.container.className = "task-dashboard-container";
    }
    /**
     * Carrega dados da API e renderiza
     */
    loadAndRender() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield this.loadProjectsAndTasks();
                yield this.render();
            }
            catch (error) {
                console.error("Erro ao carregar dados:", error);
                this.container.innerHTML = `
        <div class="empty-state">
          <p>Erro ao carregar tarefas</p>
        </div>
      `;
            }
            return this.container;
        });
    }
    /**
     * Carrega projetos e tarefas da API
     */
    loadProjectsAndTasks() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                // Carregar todos os usuários uma única vez
                const allUsers = yield UserService.getUsers();
                this.userMap.clear();
                allUsers.forEach((user) => {
                    this.userMap.set(user.getId(), user);
                });
                // Buscar todos os projetos
                const projectsData = yield ProjectService.getProjects();
                this.projects = projectsData;
                // Para cada projeto, buscar suas tarefas
                for (const project of this.projects) {
                    try {
                        const tasks = yield TaskService.getTasksByProject(project.getId());
                        this.tasksByProject.set(project.getId(), tasks);
                    }
                    catch (error) {
                        console.warn(`Erro ao buscar tarefas do projeto ${project.getId()}:`, error);
                        this.tasksByProject.set(project.getId(), []);
                    }
                }
            }
            catch (error) {
                console.error("Erro ao carregar projetos:", error);
                this.projects = [];
            }
        });
    }
    /**
     * Renderiza as tarefas em um container com cards agrupadas por projeto
     */
    render() {
        return __awaiter(this, void 0, void 0, function* () {
            this.container.innerHTML = "";
            if (this.projects.length === 0) {
                this.container.innerHTML = `
        <div class="empty-state">
          <p>Nenhum projeto encontrado</p>
        </div>
      `;
                return;
            }
            // Renderizar cada projeto com suas tarefas
            for (const project of this.projects) {
                const tasks = this.tasksByProject.get(project.getId()) || [];
                const projectSection = yield this.createProjectSection(project, tasks);
                this.container.appendChild(projectSection);
            }
        });
    }
    /**
     * Cria uma seção de projeto com suas tarefas
     */
    createProjectSection(project, tasks) {
        return __awaiter(this, void 0, void 0, function* () {
            const section = document.createElement("div");
            section.className = "project-task-section";
            // Cabeçalho do projeto
            const header = document.createElement("div");
            header.className = "project-task-header";
            header.innerHTML = `<h2>${project.getName()}</h2><div class="project-separator"></div>`;
            // Wrapper dos cards
            const tasksWrapper = document.createElement("div");
            tasksWrapper.className = "tasks-cards-wrapper";
            if (tasks.length === 0) {
                tasksWrapper.innerHTML = `
        <div class="empty-project-state">
          <p>Nenhuma tarefa neste projeto</p>
        </div>
      `;
            }
            else {
                // Renderizar cards das tarefas
                for (const task of tasks) {
                    const taskCard = yield createTaskCard(task, this.userMap);
                    tasksWrapper.appendChild(taskCard);
                }
            }
            section.appendChild(header);
            section.appendChild(tasksWrapper);
            return section;
        });
    }
}
