// m3_mini_projeto/dist/src/ui/modal/ModalTaskEditRight.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { UserService, CommentService, TaskService } from "../../services/index.js";
import { createButton, createHeadingTitle, createSection, } from "../dom/index.js";
import { setCardBorderColor, getCardBorderColor, showInfoBanner, } from "../../helpers/index.js";
import { TaskStatus } from "../../tasks/TaskStatus.js";
export function renderEditTaskRightPanel(task) {
    return __awaiter(this, void 0, void 0, function* () {
        const rightContainer = document.createElement("div");
        rightContainer.className = "edit-task-right";
        rightContainer.style.flex = "1";
        rightContainer.style.display = "flex";
        rightContainer.style.flexDirection = "column";
        const tagsHeader = createHeadingTitle("h4", "TAGS");
        tagsHeader.style.textAlign = "center";
        tagsHeader.style.marginBottom = "0.5rem";
        const tagsContainer = document.createElement("div");
        tagsContainer.style.display = "flex";
        tagsContainer.style.flexWrap = "wrap";
        tagsContainer.style.gap = "0.5rem";
        tagsContainer.style.marginBottom = "1rem";
        rightContainer.append(tagsHeader, tagsContainer);
        yield loadTaskTags(task, tagsContainer);
        const commentsHeader = createHeadingTitle("h4", "COMENTÁRIOS");
        commentsHeader.style.textAlign = "center";
        commentsHeader.style.marginBottom = "0.5rem";
        rightContainer.appendChild(commentsHeader);
        const authorSection = createSection("commentAuthorSection");
        authorSection.className = "form-group";
        authorSection.style.marginBottom = "0.5rem";
        const authorLabel = document.createElement("label");
        authorLabel.setAttribute("for", "commentAuthorSelect");
        authorLabel.textContent = "Comentário feito por";
        authorLabel.style.display = "block";
        authorLabel.style.marginBottom = "0.35rem";
        const authorSelect = document.createElement("select");
        authorSelect.id = "commentAuthorSelect";
        authorSelect.style.width = "100%";
        authorSelect.style.padding = "0.5rem";
        authorSelect.style.border = "1px solid #ccc";
        authorSelect.style.borderRadius = "4px";
        authorSection.append(authorLabel, authorSelect);
        rightContainer.appendChild(authorSection);
        const commentList = document.createElement("div");
        commentList.className = "comment-list";
        commentList.style.flex = "1";
        commentList.style.overflowY = "auto";
        commentList.style.border = "1px solid #999";
        commentList.style.padding = "0.5rem";
        commentList.style.marginBottom = "0.5rem";
        commentList.style.backgroundColor = "#e9e6e6";
        const commentInput = document.createElement("textarea");
        commentInput.rows = 3;
        commentInput.placeholder = "Escreva um comentário";
        commentInput.style.width = "100%";
        commentInput.id = "newCommentText";
        commentInput.style.marginTop = "0.5rem";
        commentInput.style.marginBottom = "0.5rem";
        commentInput.style.padding = "0.75rem";
        commentInput.style.border = "1px solid #ccc";
        commentInput.style.borderRadius = "4px";
        commentInput.style.resize = "vertical";
        const sendBtn = createButton("buttonSendComment", "Enviar Comentário", "button");
        sendBtn.style.alignSelf = "flex-start";
        sendBtn.style.backgroundColor = "#2196F3";
        sendBtn.style.color = "white";
        sendBtn.style.border = "none";
        sendBtn.style.borderRadius = "4px";
        sendBtn.style.padding = "0.65rem 1rem";
        sendBtn.style.cursor = "pointer";
        rightContainer.append(commentList, commentInput, sendBtn);
        yield populateAuthorSelect(task, authorSelect, sendBtn);
        sendBtn.onclick = () => __awaiter(this, void 0, void 0, function* () {
            const msg = commentInput.value.trim();
            const selectedAuthor = Number(authorSelect.value);
            if (!msg) {
                showInfoBanner("Escreva o comentário antes de enviar.", "error-banner");
                return;
            }
            if (!selectedAuthor || selectedAuthor <= 0) {
                showInfoBanner("Selecione um membro para assinar o comentário.", "error-banner");
                return;
            }
            try {
                const created = yield CommentService.createTaskComment(task.getId(), {
                    userId: selectedAuthor,
                    content: msg,
                });
                if (!created) {
                    throw new Error("Comentário não foi criado.");
                }
                showInfoBanner("Comentário criado com sucesso.", "success-banner");
                commentInput.value = "";
                yield refreshComments(task, commentList);
            }
            catch (error) {
                console.error("Erro ao criar comentário:", error);
                showInfoBanner("Erro ao criar comentário.", "error-banner");
            }
        });
        yield refreshComments(task, commentList);
        return rightContainer;
    });
}
function populateAuthorSelect(task, authorSelect, sendBtn) {
    return __awaiter(this, void 0, void 0, function* () {
        var _a;
        authorSelect.innerHTML = "";
        const assignees = ((_a = task.getAssignees) === null || _a === void 0 ? void 0 : _a.call(task)) || [];
        if (assignees.length === 0) {
            const placeholder = document.createElement("option");
            placeholder.value = "";
            placeholder.textContent = "Nenhum membro atribuído";
            placeholder.disabled = true;
            placeholder.selected = true;
            authorSelect.appendChild(placeholder);
            authorSelect.disabled = true;
            sendBtn.disabled = true;
            return;
        }
        const uniqueIds = Array.from(new Set(assignees.map((assignee) => assignee.user_id)));
        const users = yield Promise.all(uniqueIds.map((id) => UserService.getUserById(id)));
        uniqueIds.forEach((userId, index) => {
            const user = users[index];
            const option = document.createElement("option");
            option.value = String(userId);
            option.textContent = user ? user.getName() : `Membro ${userId}`;
            authorSelect.appendChild(option);
        });
        authorSelect.disabled = false;
        sendBtn.disabled = false;
    });
}
function loadTaskTags(task, container) {
    return __awaiter(this, void 0, void 0, function* () {
        container.innerHTML = "";
        try {
            const tags = yield TaskService.getTaskTags(task.getId());
            if (tags.length === 0) {
                const noTags = document.createElement("p");
                noTags.textContent = "Nenhuma tag associada.";
                noTags.style.margin = "0";
                noTags.style.color = "#777";
                container.appendChild(noTags);
                return;
            }
            tags.forEach((tag) => {
                const badge = document.createElement("span");
                badge.textContent = tag.name;
                badge.style.padding = "0.25rem 0.6rem";
                badge.style.backgroundColor = "#eef";
                badge.style.color = "#084";
                badge.style.borderRadius = "12px";
                badge.style.fontSize = "0.8rem";
                badge.style.border = "1px solid #cce";
                container.appendChild(badge);
            });
        }
        catch (error) {
            console.warn("Erro ao carregar tags da tarefa:", error);
            const errorMsg = document.createElement("p");
            errorMsg.textContent = "Não foi possível carregar tags.";
            errorMsg.style.margin = "0";
            errorMsg.style.color = "#e74c3c";
            container.appendChild(errorMsg);
        }
    });
}
function refreshComments(task, commentList) {
    return __awaiter(this, void 0, void 0, function* () {
        commentList.innerHTML = "";
        try {
            const comments = yield CommentService.getTaskComments(task.getId());
            if (comments.length === 0) {
                const empty = document.createElement("p");
                empty.textContent = "Nenhum comentário ainda.";
                empty.style.color = "#555";
                empty.style.margin = "0";
                commentList.appendChild(empty);
                return;
            }
            const userIds = Array.from(new Set(comments.map((comment) => comment.user_id)));
            const users = yield Promise.all(userIds.map((id) => UserService.getUserById(id)));
            const userMap = new Map();
            userIds.forEach((id, index) => {
                const user = users[index];
                userMap.set(id, user ? user.getName() : `Membro ${id}`);
            });
            comments.forEach((comment) => {
                const item = document.createElement("div");
                item.className = "comment-item";
                item.style.border = "1px solid #ccc";
                item.style.borderRadius = "8px";
                item.style.padding = "0.75rem";
                item.style.marginBottom = "0.5rem";
                item.style.backgroundColor = "#fff";
                const authorName = document.createElement("strong");
                authorName.textContent = userMap.get(comment.user_id) || "Membro";
                const createdAt = document.createElement("span");
                createdAt.style.display = "block";
                createdAt.style.fontSize = "0.8rem";
                createdAt.style.color = "#666";
                createdAt.style.marginTop = "0.25rem";
                createdAt.textContent = comment.created_at
                    ? new Date(comment.created_at).toLocaleString("pt-BR")
                    : "Data não disponível";
                const commentText = document.createElement("p");
                commentText.textContent = comment.content;
                commentText.style.margin = "0.5rem 0 0 0";
                commentText.style.whiteSpace = "pre-wrap";
                commentText.style.lineHeight = "1.4";
                item.append(authorName, createdAt, commentText);
                const borderColor = getCardBorderColor(TaskStatus.ARCHIVED);
                setCardBorderColor(item, borderColor);
                commentList.appendChild(item);
            });
        }
        catch (error) {
            console.error("Erro ao carregar comentários:", error);
            const errorMsg = document.createElement("p");
            errorMsg.textContent = "Não foi possível carregar comentários.";
            errorMsg.style.color = "#e74c3c";
            errorMsg.style.margin = "0";
            commentList.appendChild(errorMsg);
        }
    });
}
