// m3_mini_projeto/dist/src/ui/modal/ModalSprintForm.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { SprintService, ProjectStatusService } from "../../services/index.js";
import { GlobalValidators } from "../../utils/index.js";
import { loadSprintsPage } from "../sprints/SprintsPageUI.js";
import { createButton, createForm, createHeadingTitle, createInputGroup, createSection, } from "../dom/index.js";
import { showInfoBanner } from "../../helpers/index.js";
function setupSprintFormLogic(form, fields, errors, modal, projectId, sprintToEdit) {
    form.onsubmit = (e) => __awaiter(this, void 0, void 0, function* () {
        e.preventDefault();
        // Obter os valores dos campos
        const name = fields.name.value;
        const description = fields.description.value;
        const startDate = fields.startDate.value;
        const endDate = fields.endDate.value;
        // Reset de estados de erro
        errors.nameErr.textContent = "";
        errors.descriptionErr.textContent = "";
        errors.startDateErr.textContent = "";
        errors.endDateErr.textContent = "";
        let isValid = true;
        // Validações
        if (!GlobalValidators.isNonEmpty(name.trim())) {
            errors.nameErr.textContent = "O nome do sprint não pode estar vazio.";
            isValid = false;
        }
        if (!GlobalValidators.minLength(name.trim(), 3)) {
            errors.nameErr.textContent =
                "O nome do sprint deve ter pelo menos 3 caracteres.";
            isValid = false;
        }
        if (!GlobalValidators.isNonEmpty(startDate)) {
            errors.startDateErr.textContent = "A data de início é obrigatória.";
            isValid = false;
        }
        if (!GlobalValidators.isNonEmpty(endDate)) {
            errors.endDateErr.textContent = "A data de fim é obrigatória.";
            isValid = false;
        }
        if (startDate && endDate) {
            const start = new Date(startDate);
            const end = new Date(endDate);
            if (start >= end) {
                errors.endDateErr.textContent =
                    "A data de fim deve ser depois da data de início.";
                isValid = false;
            }
        }
        // Verificação Final
        if (isValid) {
            // Criar objeto do sprint (ID 0 será gerado pela base de dados)
            const sprintData = {
                id: (sprintToEdit === null || sprintToEdit === void 0 ? void 0 : sprintToEdit.id) || 0, // ID existente ou placeholder para novo
                project_id: projectId, // Projeto obtido via argumento
                name: name.trim(),
                description: description || "",
                status_id: parseInt(fields.statusId.value), // Status selecionado no dropdown
                start_date: startDate,
                end_date: endDate,
            };
            try {
                // Criar ou atualizar sprint via serviço (envia para a API)
                if (sprintToEdit) {
                    yield SprintService.updateSprint(sprintData.id, sprintData);
                    showInfoBanner(`INFO: O sprint ${name} foi atualizado com sucesso.`, "info-banner");
                }
                else {
                    yield SprintService.createSprint(sprintData);
                    showInfoBanner(`INFO: O sprint ${name} foi criado com sucesso.`, "info-banner");
                }
                // Obter todos os sprints e renderizar
                const sprints = yield SprintService.getSprints();
                yield loadSprintsPage(sprints);
                modal.remove();
            }
            catch (error) {
                const action = sprintToEdit ? "atualizar" : "criar";
                showInfoBanner(`ERRO: Não foi possível ${action} o sprint ${name}.`, "error-banner");
                console.error(`Erro ao ${action} sprint:`, error);
            }
        }
        else {
            showInfoBanner(`ERRO: O sprint não foi criado. Verifique os erros no formulário.`, "error-banner");
        }
    });
}
/**
 * Função Principal: Monta o Modal no DOM
 * @param projectId - ID do projeto para o qual criar o sprint
 * @param sprintToEdit - Sprint existente para edição (opcional). Se não fornecido, modo CREATE
 */
export function renderSprintModal(projectId, sprintToEdit) {
    return __awaiter(this, void 0, void 0, function* () {
        const modal = createSection("modalSprintForm");
        modal.classList.add("modal");
        const content = createSection("modalSprintContent");
        content.classList.add("modal-content");
        const closeBtn = document.createElement("span");
        closeBtn.classList.add("close");
        closeBtn.innerHTML = "&times;";
        closeBtn.onclick = () => modal.remove();
        const titleHeading = createHeadingTitle("h2", sprintToEdit ? "Editar Sprint" : "Adicionar Sprint");
        const form = createForm("formSprint");
        // Criação dos campos usando a função auxiliar
        const nameData = createInputGroup("Nome do Sprint", "sprintNameInput", "text", "inserir o nome do sprint");
        if (sprintToEdit) {
            nameData.input.value = sprintToEdit.name;
        }
        // Criar descrição como textarea com 4 linhas
        const descriptionGroup = document.createElement("section");
        descriptionGroup.className = "form-group";
        const descriptionLabel = document.createElement("label");
        descriptionLabel.htmlFor = "sprintDescriptionInput";
        descriptionLabel.textContent = "Descrição";
        const descriptionTextarea = document.createElement("textarea");
        descriptionTextarea.id = "sprintDescriptionInput";
        descriptionTextarea.rows = 4;
        descriptionTextarea.placeholder = "inserir a descrição do sprint (opcional)";
        if (sprintToEdit) {
            descriptionTextarea.value = sprintToEdit.description || "";
        }
        descriptionGroup.append(descriptionLabel, descriptionTextarea);
        const descriptionData = {
            section: descriptionGroup,
            input: descriptionTextarea,
            errorSection: document.createElement("section"),
        };
        descriptionData.errorSection.id = "sprintDescriptionInputError";
        descriptionData.errorSection.className = "error-message";
        descriptionGroup.append(descriptionData.errorSection);
        const startDateData = createInputGroup("Data de Início", "sprintStartDateInput", "date", "selecionar data de início");
        if (sprintToEdit) {
            startDateData.input.value = sprintToEdit.start_date;
        }
        const endDateData = createInputGroup("Data de Fim", "sprintEndDateInput", "date", "selecionar data de fim");
        if (sprintToEdit) {
            endDateData.input.value = sprintToEdit.end_date;
        }
        // Criar select de Status
        const statusGroup = document.createElement("section");
        statusGroup.className = "form-group";
        const statusLabel = document.createElement("label");
        statusLabel.htmlFor = "sprintStatusInput";
        statusLabel.textContent = "Estado";
        const statusSelect = document.createElement("select");
        statusSelect.id = "sprintStatusInput";
        // Carregar estados da API
        try {
            const statuses = yield ProjectStatusService.getProjectStatuses();
            const defaultStatusId = sprintToEdit ? sprintToEdit.status_id : 1; // Status atual ou padrão
            statuses.forEach((status) => {
                const option = document.createElement("option");
                option.value = String(status.id);
                option.textContent = status.name;
                if (status.id === defaultStatusId)
                    option.selected = true;
                statusSelect.appendChild(option);
            });
        }
        catch (error) {
            console.error("Erro ao carregar estados de projeto:", error);
        }
        statusGroup.append(statusLabel, statusSelect);
        const submitBtn = createButton("button", sprintToEdit ? "Atualizar Sprint" : "Criar Sprint", "submit");
        form.append(nameData.section, descriptionData.section, startDateData.section, endDateData.section, statusGroup, submitBtn);
        content.append(closeBtn, titleHeading, form);
        modal.append(content);
        document.body.appendChild(modal);
        // Ligar a lógica ao formulário
        setupSprintFormLogic(form, {
            name: nameData.input,
            description: descriptionData.input,
            startDate: startDateData.input,
            endDate: endDateData.input,
            statusId: statusSelect,
        }, {
            nameErr: nameData.errorSection,
            descriptionErr: descriptionData.errorSection,
            startDateErr: startDateData.errorSection,
            endDateErr: endDateData.errorSection,
        }, modal, projectId, sprintToEdit);
        // Fechar ao clicar fora
        modal.onclick = (e) => {
            if (e.target === modal)
                modal.remove();
        };
        modal.style.display = "block";
    });
}
