// m3_mini_projeto/dist/src/ui/modal/ModalTaskForm.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { GlobalValidators } from "../../utils/index.js";
import { TaskService } from "../../services/index.js";
import { createButton, createForm, createHeadingTitle, createInputGroup, createSection, createSelectGroup, } from "../dom/index.js";
import { showInfoBanner } from "../../helpers/index.js";
function setupTaskFormLogic(form, fields, errors, modal, projectId, taskToEdit, user, onSuccess) {
    form.onsubmit = (e) => __awaiter(this, void 0, void 0, function* () {
        var _a, _b;
        e.preventDefault();
        const title = fields.title.value;
        const category = fields.category.value;
        const type = fields.type.value;
        // Reset de estados
        errors.titleErr.textContent = "";
        errors.categoryErr.textContent = "";
        errors.typeErr.textContent = "";
        let isValid = true;
        if (!GlobalValidators.minLength(title.trim(), 3)) {
            errors.titleErr.textContent = "O titulo deve ter pelo menos 3 caracteres.";
            isValid = false;
        }
        if (!GlobalValidators.isNonEmpty(title.trim())) {
            errors.titleErr.textContent = "O titulo não pode estar vazio.";
            isValid = false;
        }
        if (!GlobalValidators.isNonEmpty(category.trim())) {
            errors.categoryErr.textContent = "A categoria não pode estar vazia.";
            isValid = false;
        }
        if (!GlobalValidators.isNonEmpty(type.trim())) {
            errors.typeErr.textContent = "O tipo não pode estar vazio.";
            isValid = false;
        }
        if (isValid) {
            const taskId = (_a = taskToEdit === null || taskToEdit === void 0 ? void 0 : taskToEdit.id) !== null && _a !== void 0 ? _a : (_b = taskToEdit === null || taskToEdit === void 0 ? void 0 : taskToEdit.getId) === null || _b === void 0 ? void 0 : _b.call(taskToEdit);
            const taskData = {
                id: taskId || 0,
                project_id: projectId,
                title: title.trim(),
                category: category.trim(),
                type: type.trim(),
            };
            try {
                if (taskToEdit) {
                    yield TaskService.updateTask(taskId, taskData);
                    showInfoBanner(`INFO: A tarefa "${title}" foi atualizada com sucesso.`, "success-banner");
                }
                else {
                    yield TaskService.createTask(taskData);
                    showInfoBanner(`INFO: A tarefa "${title}" foi criada com sucesso.`, "success-banner");
                }
                modal.remove();
                if (onSuccess)
                    yield onSuccess();
            }
            catch (error) {
                const action = taskToEdit ? "atualizar" : "criar";
                showInfoBanner(`ERRO: Não foi possível ${action} a tarefa.`, "error-banner");
                console.error(`Erro ao ${action} tarefa:`, error);
            }
        }
        else {
            showInfoBanner(`ERRO: Verifique os erros no formulário.`, "error-banner");
        }
    });
}
/**
 *  Função Principal: Monta o Modal no DOM
 * @param projectId - ID do projeto ao qual a tarefa pertence (obrigatório)
 * @param taskToEdit - Tarefa existente para edição (opcional). Se não fornecido, modo CREATE
 * @param user - Utilizador atual (opcional)
 */
export function renderTaskModal(projectId, taskToEdit, user, onSuccess) {
    return __awaiter(this, void 0, void 0, function* () {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
        const modal = createSection("modalTaskForm");
        modal.classList.add("modal");
        const content = createSection("modalTaskContent");
        content.classList.add("modal-content");
        const closeBtn = document.createElement("span");
        closeBtn.classList.add("close");
        closeBtn.innerHTML = "&times;";
        closeBtn.onclick = () => modal.remove();
        const titleHeading = createHeadingTitle("h2", taskToEdit ? "Editar Tarefa" : "Adicionar Tarefa");
        const form = createForm("formTask");
        // Criação dos campos usando a função auxiliar
        const titleData = createInputGroup("Titulo", "taskTitleInput", "text", "inserir o titulo da tarefa");
        const titleValue = (_c = (_a = taskToEdit === null || taskToEdit === void 0 ? void 0 : taskToEdit.title) !== null && _a !== void 0 ? _a : (_b = taskToEdit === null || taskToEdit === void 0 ? void 0 : taskToEdit.getTitle) === null || _b === void 0 ? void 0 : _b.call(taskToEdit)) !== null && _c !== void 0 ? _c : "";
        if (titleValue) {
            titleData.input.value = titleValue;
        }
        const taskCategory = ["Trabalho", "Pessoal", "Estudo"];
        const categoryData = createSelectGroup("Cargo", "categoryID", taskCategory);
        const categoryValue = (_g = (_d = taskToEdit === null || taskToEdit === void 0 ? void 0 : taskToEdit.category) !== null && _d !== void 0 ? _d : (_f = (_e = taskToEdit === null || taskToEdit === void 0 ? void 0 : taskToEdit.getTaskCategory) === null || _e === void 0 ? void 0 : _e.call(taskToEdit)) === null || _f === void 0 ? void 0 : _f.name) !== null && _g !== void 0 ? _g : "";
        if (categoryValue) {
            categoryData.select.value = categoryValue;
        }
        const taskType = ["Bugs", "Feature", "Task"];
        const TypeData = createSelectGroup("Tipo", "typeID", taskType);
        const typeValue = (_k = (_h = taskToEdit === null || taskToEdit === void 0 ? void 0 : taskToEdit.type) !== null && _h !== void 0 ? _h : (_j = taskToEdit === null || taskToEdit === void 0 ? void 0 : taskToEdit.getType) === null || _j === void 0 ? void 0 : _j.call(taskToEdit)) !== null && _k !== void 0 ? _k : "";
        if (typeValue) {
            TypeData.select.value = typeValue;
        }
        const submitBtn = createButton("button", taskToEdit ? "Atualizar" : "Adicionar", "submit");
        form.append(titleData.section, categoryData.section, TypeData.section, submitBtn);
        content.append(closeBtn, titleHeading, form);
        modal.append(content);
        document.body.appendChild(modal);
        // Ligar a lógica ao formulário
        setupTaskFormLogic(form, {
            title: titleData.input,
            category: categoryData.select,
            type: TypeData.select,
        }, {
            titleErr: titleData.errorSection,
            categoryErr: categoryData.errorSection,
            typeErr: TypeData.errorSection,
        }, modal, projectId, taskToEdit, user, onSuccess);
        // Fechar ao clicar fora
        modal.onclick = (e) => {
            if (e.target === modal)
                modal.remove();
        };
        modal.style.display = "block";
    });
}
