// m3_mini_projeto/dist/src/services/assignmentService.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import * as fetchTaskAssignees from "../api/fetchTaskAssignees.js";
/* Serviço para gerir a atribuição de tarefas a utilizadores */
export class AssignmentService {
    /* Obtém a lista de atribuições de tarefas da API */
    static getTaskAssignees() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield fetchTaskAssignees.getTaskAssignees();
        });
    }
    /* Obtém uma atribuição de tarefa por ID da API */
    static getTaskAssigneeById(id) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield fetchTaskAssignees.getTaskAssigneeById(id);
        });
    }
    /* Cria uma nova atribuição de tarefa na API */
    static createTaskAssignee(assignee) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield fetchTaskAssignees.createTaskAssignee(assignee);
        });
    }
    /* Atualiza uma atribuição de tarefa na API */
    static updateTaskAssignee(id, assignee) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield fetchTaskAssignees.updateTaskAssignee(id, assignee);
        });
    }
    /* Exclui uma atribuição de tarefa na API */
    static deleteTaskAssignee(id) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield fetchTaskAssignees.deleteTaskAssignee(id);
        });
    }
}
