// m3_mini_projeto/dist/src/services/sprintService.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import * as fetchSprints from "../api/fetchSprints.js";
/* Serviço para gerenciar sprints */
export class SprintService {
    /* Função para obter a lista de sprints */
    static getSprints(sort, search) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield fetchSprints.getSprints(sort, search);
        });
    }
    /* Função para obter um sprint por ID */
    static getSprintById(id) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield fetchSprints.getSprintById(id);
        });
    }
    /* Função para criar um novo sprint */
    static createSprint(sprint) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield fetchSprints.createSprint(sprint);
        });
    }
    /* Função para atualizar um sprint existente */
    static updateSprint(id, sprint) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield fetchSprints.updateSprint(id, sprint);
        });
    }
    /* Função para excluir um sprint */
    static deleteSprint(id) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield fetchSprints.deleteSprint(id);
        });
    }
}
