// m3_mini_projeto/dist/src/services/sprintTaskService.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import * as fetchSprintTasks from "../api/index.js";
/* Serviço para gerenciar tarefas de sprint */
export class SprintTaskService {
    /* Função para obter a lista de tarefas de sprint */
    static getSprintTasks() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield fetchSprintTasks.getSprintTasks();
        });
    }
    /* Função para obter uma tarefa de sprint por ID */
    static getSprintTaskById(id) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield fetchSprintTasks.getSprintTaskById(id);
        });
    }
    /* Função para criar uma nova tarefa de sprint */
    static createSprintTask(sprintTask) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield fetchSprintTasks.createSprintTask(sprintTask);
        });
    }
    /* Função para atualizar uma tarefa de sprint existente */
    static updateSprintTask(id, sprintTask) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield fetchSprintTasks.updateSprintTask(id, sprintTask);
        });
    }
    /* Função para excluir uma tarefa de sprint */
    static deleteSprintTask(id) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield fetchSprintTasks.deleteSprintTask(id);
        });
    }
}
