// m3_mini_projeto/dist/src/services/teamService.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import * as fetchTeams from "../api/fetchTeams.js";
/* Serviço para gerenciar equipes */
export class TeamService {
    /* Função para obter a lista de equipes */
    static getTeams() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield fetchTeams.getTeams();
        });
    }
    /* Função para obter uma equipe por ID */
    static getTeamById(id) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield fetchTeams.getTeamById(id);
        });
    }
    /* Função para criar uma nova equipe */
    static createTeam(team) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield fetchTeams.createTeam(team);
        });
    }
    /* Função para atualizar uma equipe existente */
    static updateTeam(id, team) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield fetchTeams.updateTeam(id, team);
        });
    }
    /* Função para excluir uma equipe */
    static deleteTeam(id) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield fetchTeams.deleteTeam(id);
        });
    }
}
