// m3_mini_projeto/dist/src/services/teamMemberService.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import * as fetchTeamMembers from "../api/fetchTeamMembers.js";
/* Serviço para gerenciar membros de equipe */
export class TeamMemberService {
    /* Função para obter a lista de membros de equipe */
    static getTeamMembers() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield fetchTeamMembers.getTeamMembers();
        });
    }
    /* Função para obter um membro de equipe por ID */
    static getTeamMemberById(id) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield fetchTeamMembers.getTeamMemberById(id);
        });
    }
    /* Função para criar um novo membro de equipe */
    static createTeamMember(member) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield fetchTeamMembers.createTeamMember(member);
        });
    }
    /* Função para atualizar um membro de equipe existente */
    static updateTeamMember(id, member) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield fetchTeamMembers.updateTeamMember(id, member);
        });
    }
    /* Função para excluir um membro de equipe */
    static deleteTeamMember(id) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield fetchTeamMembers.deleteTeamMember(id);
        });
    }
}
